### DISCUSSION TITLE
Anyone interested to bring-up test with LIBXSMM on ARM?

### DISCUSSION THREAD

**Discussion by hfp** [cp2k/cp2k#2602] [Team Posts]
We have worked out [AArch64 support in LIBXSMM](https://github.com/hfp/libxsmm/wiki/Compatibility#arm-aarch64) covering baseline AArch64 (v8.1) as well as started to work on SVE (A64FX). Similar to x86, we are only covering 64-bit platforms. Since, [CP2K Dashboard](https://dashboard.cp2k.org/) just covers 32-bit ARM, I was wondering if 64-bit ARM is desired?

With respect to LIBXSMM, SMM kernels (and some other stuff) are supported and should be able to cover the entire CP2K application. We still build up and polish the support for our next release (v1.17), i.e., LIBXSMM/master is currently needed. If somebody wants to setup an ARM CI-test for CP2K, I can share a stable/validated/fixed SHA of LIBXSMM/master.

**Comment by oschuett** [comment]
Yes, we should pay attention to 64-bit ARM. As a first step we could try the [Raspberry-Pi-OS-64bit](https://github.com/raspberrypi/Raspberry-Pi-OS-64bit), which is currently in beta. As a next step we could look into cloud offers. However, I don't think it's feasible to optimize for the architecture until HPC-worthy chips, like Ampere Altra, become widely available.

**Comment by hfp** [comment]
( As a clarifying note, for us the motivation is to have leadership HPC hardware [even without software differentiation] ).

I have no experience with Raspberry-Pi-OS-64bit, but I can help if anything breaks (or does not work out of the box). We use an NVidia Jetson system (AGX) for our CI-tests, which is 64-bit out of the box. We may get our M1 based system up as soon as we can access our lab with reasonable measure (pandemic). However, we have not invested yet to check with anything more HPC centric.

**Comment by dev-zero** [comment]
I can see what could be done on Isambard2 since they have CP2K In their performance test suite already.
What would be interesting alongside with an ARM platform with SVE are also the compilers on those machines:

* armflang is almost there for DBCSR, less so for CP2K, but they are working on it
* Fujitsu can build DBCSR with minimal modifications and is capable of providing very good performance already on a single node

**Comment by hfp** [comment]
I can (re-)make own connection with Bristol folks or perhaps Fujitsu directly, but I would prefer if somebody can help/introduce me. I would probably ask for two logins/accounts for the sake of SVE development (me plus a colleague). There is no need for compute time or at least not big resources (perhaps portal/login node is enough if SVE is present).

**Comment by dev-zero** [comment]
Ok, Isambard2 NHs for CP2K regtesting approved. @hfp wrt account, please check your mail ;-)

**Comment by mkrack** [comment]
libxsmm compiles with `make PLATFORM=1 FC=gfortran CC=gcc CXX=g++ test` on a Raspberry PI4 running `armv7l` using GNU 8.3.0. All tests are OK. The current [ARMv7l regression tester](https://dashboard.cp2k.org/archive/arm-gnu-psmp/index.html) has libxsmm disabled, however, because an error is thrown by `libxsmm_macros.h` when I try to compile CP2K.

Update: After I added the flag `-DLIBXSMM_PLATFORM_FORCE` in the arch file, the compilation continues.

**Comment by hfp** [comment]
The `PLATFORM=1` is mainly to unlock compilation everywhere like ARM 32-bit, etc. It only makes the code not based on JIT compilation available (and even that is not supported; we only want to test compilation everywhere even if the platform is not supported). This toggle was also introduced to intentionally produce a compile-time error by default (if  `PLATFORM=1` is not set) like : `LIBXSMM requires X86_64, AArch64, or a compatible CPU!` or `LIBXSMM is only supported on 64-bit platforms!`. These errors help us to avoid LIBXSMM gets upstreamed for unsupported platforms.

What we worked on is to produce native AArch64 code on compatible CPUs at runtime (JIT). If you force compilation on an unsupported platform (`PLATFORM=1`), functionality that is normally returning code at runtime just returns a NULL-pointer. On supported platforms JIT code generation always succeeds and never returns a NULL-pointer. CP2K still (legacy) tests for valid code (non-NULL) and hence always falls back in such case (to GEMM for instance).

**Comment by mkrack** [comment]
@hfp Thanks for the explanation. OK, then it still does not make sense to include libxsmm with armv7l.
